### ISSUE TITLE
System.Security.SecurityException: Request for the permission of type 'System.Net.WebPermission, System, Version=4.0.0.0' failed."

### ISSUE THREAD

**Issue by rexkenley** [restsharp/RestSharp#1091]
Hi

I am using RestSharp to call web rest services from within crm365. However, I am getting this error whenever I made the call.

System.Security.SecurityException: Request for the permission of type 'System.Net.WebPermission, System, Version=4.0.0.0' failed."

What must I do to remedy this?

Ps.

I also noticed that the new version seems to have issues with ILMerge with MS Tuple library.

**Comment by mdach** [comment]
Hello,

I was looking around online for this exception, turns out you would need to set your trust level to "full" in your web.config file. Try it and see if that works!

**Comment by alexeyzimarev** [comment]
This is not a RestSharp issue. Second link in Google is exactly about Dynamics: https://community.dynamics.com/crm/f/117/t/217128

**Comment by rexkenley** [comment]
2nd link in google doesn't resolve the issue.

Why does RestSharp need System.Net.WebPermission? I created a HttpWebRequest that builds of the IRestClient.BuildUri and it works fine.

**Comment by alexeyzimarev** [comment]
I am sorry, I might be wrong of course. But there is no way I can help you here. You are calling a particular system that apparently results in this exception but there are no other cases reported with similar symptoms. I am not able to reproduce simply because I am not using your system. If you are able to create a repro repository, I will be happy to at least try helping.

**Comment by alexeyzimarev** [comment]
Plus, the stacktrace is missing in your issue.

**Comment by rexkenley** [comment]
Alex

Thanks for taking the time to look at this, much appreciate it. 

This is the environment I am running RestSharp in, do any of these limitations apply? The environment is online, so there is the added restriction that I cannot change the registry.

- Only the HTTP and HTTPS protocols are allowed.
- Access to localhost (loopback) is not permitted.
- IP addresses cannot be used. You must use a named web address that requires DNS name resolution.
- Anonymous authentication is supported and recommended. There is no provision for prompting the logged on user for credentials or saving those credentials.

https://msdn.microsoft.com/en-us/library/gg334752.aspx

Here is the stack trace. 

Request for the permission of type 'System.Net.WebPermission, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' failed.
at System.Security.CodeAccessSecurityEngine.Check(Object demand, StackCrawlMark& stackMark, Boolean isPermSet)
at System.Security.CodeAccessPermission.Demand()
at System.Net.HttpWebRequest.set_Proxy(IWebProxy value)
at RestSharp.Http.ConfigureWebRequest(String method, Uri url)
at RestSharp.Http.PostPutInternal(String method)
at RestSharp.Http.AsPost(String httpMethod)
at RestSharp.RestClient.DoExecuteAsPost(IHttp http, String method)
at RestSharp.RestClient.Execute(IRestRequest request, String httpMethod, Func`3 getResponse)

"type": "System.ServiceModel.FaultException`1[[Microsoft.Xrm.Sdk.OrganizationServiceFault, Microsoft.Xrm.Sdk, Version=8.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]]",
"stacktrace": "   
at Microsoft.Crm.Extensibility.OrganizationSdkServiceInternal.Execute(OrganizationRequest request, CorrelationToken correlationToken, CallerOriginToken callerOriginToken, WebServiceType serviceType, Boolean checkAdminMode, ExecutionContext executionContext)   
at Microsoft.Crm.Extensibility.OData.CrmODataExecutionContext.Execute(OrganizationRequest request, ExecutionContext executionContext)
at Microsoft.Crm.Extensibility.OData.CrmODataServiceDataProvider.ExecuteOperation(CrmODataExecutionContext context, EdmOperation edmOperation, Dictionary`2 parameters, Dictionary`2 boundParameters)
at Microsoft.Crm.Extensibility.OData.ActionController.ProcessOperationRequest(String operationName, Dictionary`2 operationParameters, EntityReference entityReference, String boundEntityName, String boundEntityType)
at Microsoft.Crm.Extensibility.OData.ActionController.PostUnboundAction(String operationName, ODataUntypedActionParameters parameters)
at lambda_method(Closure , Object , Object[] )
at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.<>c__DisplayClass10.<GetExecutor>b__9(Object instance, Object[] methodParameters)
at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ActionExecutor.Execute(Object instance, Object[] arguments)
at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync(HttpControllerContext controllerContext, IDictionary`2 arguments, CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
t System.Web.Http.Controllers.ApiControllerActionInvoker.<InvokeActionAsyncCore>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
at System.Web.Http.Controllers.ActionFilterResult.<ExecuteAsync>d__2.MoveNext()
End of stack trace from previous location where exception was thrown ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
at System.Web.Http.Dispatcher.HttpControllerDispatcher.<SendAsync>d__1.MoveNext()"

**Comment by alexeyzimarev** [comment]
This is exactly what discussion in the link I sent you went to. The code is unable to read the proxy settings from web.config in Dynamics. Presumably this is your embedded code.

**Comment by rexkenley** [comment]
Is there a way to not use the proxy settings if it's inaccessible? As I have stated previously I can use HttpWebRequest with the url created by RestSharp with no issue.

Or is this the only way to do it?

**Comment by alexeyzimarev** [comment]
If the Proxy property in the client is not set, we try to get the default proxy. You can specify _no proxy_ by using GlobalProxySelection.GetEmptyWebProxy

**Comment by rexkenley** [comment]
I tried that and several other settings and none of them worked.

client.Proxy = GlobalProxySelection.GetEmptyWebProxy *obsolete
client.Proxy = WebRequest.GetSystemWebProxy();
client.Proxy = WebRequest.DefaultWebProxy;
client.Proxy = HttpWebRequest.GetSystemWebProxy();

Also I am using package id="RestSharp" version="106.1.0" targetFramework="net452". 

I am NOT using the latest version because ILMerge seems to have an issue with locating the ValueTuple library.

**Comment by rexkenley** [comment]
Alex,

Is there anything else you want me to try?

**Comment by alexeyzimarev** [comment]
I can't say. I have no system to reproduce this. I just read the link that I send you and followed the stack trace. Clearly, we try to get the proxy and fail due to lack of access to web.config.

**Comment by alexeyzimarev** [comment]
You can try debugging with RestSharp source code yourself to see why is it not using the pre-assigned proxy.

**Comment by rexkenley** [comment]
Ok, thanks for your help.

**Comment by AndrewMorris-2** [comment]
@rexkenley, any luck in getting RestSharp to work in a CRM plugin? I went so far as creating a WebRequest, then borrowing the proxy off of that object. No Dice. 
I've given up on getting RestSharp to work for now, but would love to get to the bottom of this error!

I think @alexeyzimarev hit the issue right on the nose with his statement "Clearly, we try to get the proxy and fail due to lack of access to web.config". I'm guessing your plugin is running in sandbox mode, so it has no access to the filesystem. https://community.dynamics.com/crm/f/117/t/145118

With that said, see the below examples. 
I am surprise the following code works (by "code works" I mean the request is sent successfully)): 
```
var request = WebRequest.Create(uri) as HttpWebRequest;     
request.Method = "GET";     
request.KeepAlive = false;     
request.ContentType = "application/json";     
using (var response = (HttpWebResponse)request.GetResponse())       
{       
    tracer.Trace((response).StatusDescription);       
    using (Stream dataStream = response.GetResponseStream())      
    {      
        using (StreamReader reader = new StreamReader(dataStream))     
        {      
            var responseFromServer = reader.ReadToEnd();     
            data = JsonHelper.Deserialize<Data>(responseFromServer);     
        }    
    }     
}    
```

But the following does _not_ work (to be clear, by "not work", I mean I am getting the error stated at the beginning of this thread) : 
```
var uriBuilder = new UriBuilder (url);
Uri uri = uriBuilder.Uri;

var tempRequest = WebRequest.Create(url) as HttpWebRequest;
var client = new RestClient(uri.GetLeftPart(UriPartial.Authority));
client.Proxy = tempRequest.Proxy;
var request = new RestRequest(uri.PathAndQuery, Method.GET, DataFormat.Json);

IRestResponse<Data> response = client.Execute<Data>(request);

if (response.IsSuccessful)
{
    data = response.Data;
}
```

Let me know if this provides any further insight into this issue.

**Comment by rexkenley** [comment]
@AndrewMorris-2

I haven't worked on this for a while and if I remember correctly I just used an older version of RestSharp that didn't have the tuple library.

**Comment by AlexanderStruck** [comment]
I had the same issue with Dynamics and resolved it with using version 105.0.

**Comment by PedroTadeuOliveira** [comment]
Thanks @AlexanderStruck , it works just fine.
